### DISCUSSION TITLE
lfo library improvements

### DISCUSSION THREAD

**Discussion by dndrks** [monome/norns#1687] [General]
just wanted to strike up a lil' space for proposed improvements to the [lfo library](https://monome.org/docs/norns/reference/lib/lfo) :)

some ideas being passed around with `@sonocircuit`:

- add `saw` behavior (and rename current `saw` to `tri`, that took a while to catch!)
- add `phase` offset
- add `baseline` to [the constructor](https://github.com/monome/norns/blob/c5adf1cb64184c48c429c7b6c9fb9e90f2833d86/lua/lib/lfo.lua#L30C1-L30C1)
- spin up some examples (+ maybe in-lib templates/presets) to demonstrate adding lfo movement using a parameter's current value as the baseline

@sixolet , i'm curious if `toolkit`'s brilliant lifecycle has inspired any new ideas for this scripting library?

**Comment by dndrks** [comment]
addressed by https://github.com/monome/norns/pull/1692

**Reply by sonocircuit** [reply]
Hi Dan. Sorry for the long wait.

I made some minor changes which you can find [here](https://github.com/sonocircuit/lfo_test.). You can comment out between norns' `lfo.lua` and `lib/lfo.lua` in the main file to see the different behaviour.

- currently adding positive/negative offset simply clamps the lfo instead of shifting it like a DC offset (like the one you showed me from m4l). I made some changes in the `process_lfo(id)` function to address this. test scenario: set depth to 40% and the lfo will make the param oscillate between 0 and 40, add 20% offset and the param will oscillate between 20 and 60. before it clamped at 40. Ideally the offset parameter should change depending on the `baseline` settings. +100 range when set to `min`, -100 when set to `max` and +/-50 when set to `center`.
- another thing I noticed is that when norns is clocked via midi, free running lfos jump all over the place if there is midi jitter. this is because `clock.get_beats_sec()` is being called to calc the phase and slight changes at 96ppqn makes the phase constantly shift. I added some "smoothing" to prevent this but there's prob a more elegant way to do this or even address the jitter somewhere else.
- and last: under `lfo_params_visibility(state, i)`, `params[state](params, "lfo_offset_"..i)` is missing.

please let me know if something is unclear.

**Reply by dndrks** [reply]
@sonocircuit, no worries at all, thank you for this great cleanup!! the re-orientation of `offset` here make total sense, though i ran into a pocket that could use final call from ya!

using your great example script:
- set lfo min to 0.0
- set lfo max to 27.0
- set lfo depth to 11%

if i set lfo offset to 100%, then i get values between 27 and nearly 30 -- since i set my max to 27, i would expect that max to be factored into the scaling of the offset value. same with negative offsets -- i wouldn't want to go below my minimum value, but even -1% offset pushes me below 0.

some of this behavior could be worked around with the range adjustments you mentioned ("Ideally the offset parameter should change depending on the baseline settings."), but i'm curious if we should just clamp the output so that it can never go below the stated min / above the stated max, regardless of offsets and depth?

**Reply by sonocircuit** [reply]
@dndrks ah right I can see that. a few thoughts:
- the actual value doesn't go below 0 as it's clamped via the parameter's `min` value. However, the lfo _does_ display negative scaled values.
- I think it's important to keep a good balance between simplicity and flexibility. Personally I wouldn't surface the parameters `lfo_min`, `lfo_max` (and probably `baseline`) to the end user, as the same effects can be achieved with `depth`, `offset` and `phase`. If the script dev sets the `lfo_min`, `lfo_max` values to the `min/max` values of the modulated param the case you described above would not occur and the same lfo behaviour can be set with `depth` and `offset` alone. I might be overseeing specific use cases though.
- However, if you feel like keeping `lfo_min` and `lfo_max` in the params menu, I'd say preferably both the offset range should be set according to the baseline _and_ clamping min/max output values should happen if we want to ensure that the values don't go out of range.

(_I added the offset range scaling to the repo linked above. I'd be curious to know how you would implement it. I'm pretty sure I overlooked something_ )

**Reply by dndrks** [reply]
- actual value: ah, right, yes, thank you for reminder!
- you raise a super good point, though i think simplicity all depends on the situation. eg. "i want to use a square wave to flip this parameter between 19 and 46". i'd just set `min` to 19, `max` to 46, and `depth` to 100. or i could set `offset` to 19 to get my baseline, then `depth` to 27 (in order to get the scaled value up to 46), which is doable but requires more mental math than i typically want to exercise. having both avenues available helps cover a wider range of uses, so i'd be down for it all to co-exist :)
- since parameters clamp the actual value anyway, i think your changes for the `offset` range being dependent on `baseline` makes total sense + feels complete!

in summary: this rips, @sonocircuit ! would you like to submit a PR to the main codebase so your contribution is appropriately annotated in the historical view + add your name to the credits at the top of the file? <33

**Reply by sonocircuit** [reply]
- _simplicity all depends on the situation_: absolutely. that would be one of my overseen use cases.
- if the code changes are good on your behalf I'd love to do so. I'm assuming standard procedure would be. fork norns > edit lfo.lua > submit PR?

**Reply by dndrks** [reply]
code changes look good to me, and are testing well on my side! we can roll 'em into the beta and get some more testing done, though :) stoked for you to get attributed!! <3

that's the process, yeh! maybe reference this discussion in the PR for context?

**Reply by sonocircuit** [reply]
thanks dan! I'll do so. best sacha

**Comment by dndrks** [comment]
this got sorted really really well, imo, so closing the convo -- @sonocircuit feel free to reopen w/ further q's / ideas!! thank you so much for the contributions + conversations!!
